src/Shared/Targets/Wrappers/AspNetBufferingTargetWrapper.cs

```diff
@@ -0,0 +1,305 @@
+            {
+                return null;
+            }
+
+            lock (bufferDictionary)
+            {
+                if (!bufferDictionary.TryGetValue(this, out var buffer))
+                {
+                    buffer = new Internal.LogEventInfoBuffer(BufferSize, GrowBufferAsNeeded, BufferGrowLimit);
+                    bufferDictionary.Add(this, buffer);
+                }
+
+                return buffer;
+            }
+        }
+
+        /// <summary>
+        /// The Key for the buffer dictionary in the HttpContext.Items collection
+        /// </summary>
+        private static readonly object HttpContextItemsKey = new object();
+
+        /// <summary>
+        /// Return the buffer dictionary from the HttpContext.Items using HttpContextItemsKey
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        private static Dictionary<AspNetBufferingTargetWrapper, Internal.LogEventInfoBuffer> GetBufferDictionary(HttpContextBase context)
+        {
+            return context?.Items?[HttpContextItemsKey] as
+                Dictionary<AspNetBufferingTargetWrapper, Internal.LogEventInfoBuffer>;
+        }
+
+        /// <summary>
+        /// Create the buffer dictionary in the HttpContext.Items using HttpContextItemsKey
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        private static void SetBufferDictionary(HttpContextBase context)
+        {
+            context.Items[HttpContextItemsKey] = new Dictionary<AspNetBufferingTargetWrapper, Internal.LogEventInfoBuffer>();
+        }
+
+        /// <summary>
+        /// Adds the specified log event to the buffer.
+        /// NOTE: if Write is never called, this instance will not be registered in the buffer dictionary in HttpContext.Items.
+        /// That is expected normal behavior.
+        /// </summary>
+        /// <param name="logEvent">The log event.</param>
+        protected override void Write(AsyncLogEventInfo logEvent)
+        {
+            var buffer = GetOrCreateRequestBuffer(HttpContextAccessor.HttpContext);
+            if (buffer != null)
+            {
+                WrappedTarget?.PrecalculateVolatileLayouts(logEvent.LogEvent);
+                InternalLogger.Trace("Appending log event {0} to ASP.NET request buffer.", logEvent.LogEvent.SequenceID);
+                buffer.Append(logEvent);
+            }
+            else
+            {
+                InternalLogger.Trace("ASP.NET request buffer does not exist. Passing to wrapped target.");
+                WrappedTarget?.WriteAsyncLogEvent(logEvent);
+            }
+        }
+
+        /// <summary>
+        /// Called from the HttpModule or Middleware upon the end of the HTTP pipeline
+        /// Flushes all instances of this class registered in the HttpContext
+        /// </summary>
+        /// <param name="context"></param>
+        internal static void Flush(HttpContextBase context)
```

Comment:
Could this be renamed to `RequestCompleted` ?

Reply:
Completed